Add mention of module/block scope in closures (#17722)
mdn/content: files/en-us/web/javascript/closures/index.md

```js
function init() {
  var name = 'Mozilla'; // name is a local variable created by init
  function displayName() {
    // displayName() is the inner function, a closure

>     alert(name); // use variable declared in the parent function

Reviewer: Let's avoid `alert`:
```suggestion
    console.log(name); // use variable declared in the parent function
```

`init()` creates a local variable called `name` and a function called `displayName()`. The `displayName()` function is an inner function that is defined inside `init()` and is available only within the body of the `init()` function. Note that the `displayName()` function has no local variables of its own. However, since inner functions have access to the variables of outer functions, `displayName()` can access the variable `name` declared in the parent function, `init()`.

> Run the code using [this JSFiddle link](https://jsfiddle.net/xAFs9/3/) and notice that the `alert()` statement within the `displayName()` function successfully displays the value of the `name` variable, which is declared in its parent function. This is an example of _lexical scoping_, which describes how a parser resolves variable names when functions are nested. The word _lexical_ refers to the fact that lexical scoping uses the location where a variable is declared within the source code to determine where that variable is available. Nested functions have access to variables declared in their outer scope.

Reviewer: ```suggestion
Run the code using [this JSFiddle link](https://jsfiddle.net/3dxck52m/) and notice that the `console.log()` statement within the `displayName()` function successfully displays the value of the `name` variable, which is declared in its parent function. This is an example of _lexical scoping_, which describes how a parser resolves variable names when functions are nested. The word _lexical_ refers to the fact that lexical scoping uses the location where a variable is declared within the source code to determine where that variable is available. Nested functions have access to variables declared in their outer scope.
```

(We really should convert jsfiddle into internal examples (Live samples or interactive samples).

Author: True, we would do more systematic work in the future on this, I hope?

In this particular example, the scope is called a _function scope_, because the variable is accessible and only accessible within the function body where it's declared.

### Scoping with let and const

Traditionally (before ES6), JavaScript only had two kinds of scopes: _function scope_ and _global scope_. Variables declared with `var` are either function-scoped or global-scoped, depending on whether they are declared within a function or outside a function. This can be tricky, because blocks with curly braces do not create scopes:

```js
if (Math.random() > 0.5) {
  var x = 1;
} else {
  var x = 2;
}
console.log(x);
```

For people from other languages (e.g. C, Java) where blocks create scopes, the above code should throw an error on the `console.log` line, because we are outside the scope of `x` in either block. However, because blocks don't create scopes for `var`, the `var` statements here actually create a global variable. There is also [a practical example](#creating_closures_in_loops_a_common_mistake) introduced below that illustrates how this can cause actual bugs when combined with closures.

> In ES6, JavaScript introduced the `let` and `const` declarations, which, among other things like [temporary dead zones](/en-US/docs/Web/JavaScript/Reference/Statements/let#temporal_dead_zone_tdz), allow you to create block-scoped variables.

Reviewer: ```suggestion
In ES6, JavaScript introduced the `let` and `const` declarations, which, among other things like [temporal dead zones](/en-US/docs/Web/JavaScript/Reference/Statements/let#temporal_dead_zone_tdz), allow you to create block-scoped variables.
```

```js
function makeSizer(size) {
  return function () {

>     document.body.style.fontSize = size + 'px';

Reviewer: What about:
```suggestion
    document.body.style.fontSize = `${size}px`;
```

```js
// global scope
const e = 10;
function sum(a) {
  return function (b) {
    return function (c) {
      // outer functions scope
      return function (d) {
        // local scope
        return a + b + c + d + e;
      };
    };
  };
}

console.log(sum(1)(2)(3)(4)); // log 20
```

You can also write without anonymous functions:

```js
// global scope
const e = 10;
function sum(a) {
  return function sum2(b) {
    return function sum3(c) {
      // outer functions scope
      return function sum4(d) {
        // local scope
        return a + b + c + d + e;
      };
    };
  };
}

const sum2 = sum(1);
const sum3 = sum2(2);
const sum4 = sum3(3);
const result = sum4(4);
console.log(result); //log 20
```

In the example above, there's a series of nested functions, all of which have access to the outer functions' scope. In this context, we can say that closures have access to _all_ outer function scopes.

Closures can capture variables in block scopes and module scopes as well. For example, the following creates a closure over the block-scoped variable `y`:

```js
function outer() {
  const x = 5;

>   if (true) {

Reviewer: `if (true)` is an idiom to avoid. (I remember missing a top mark at uni because I forgot one)

Would this work?
```suggestion
  {
```

Author: I was trying to make this more non-trivial... I would go with `Math.random() > 0.5`, then.

Every closure has three scopes:

> - Local Scope (Own scope)

Reviewer: ```suggestion
- Local scope (Own scope)
```

> - Enclosing Scope (can be block, function, or module scope)
> - Global Scope

Reviewer: ```suggestion
- Enclosing scope (can be block, function, or module scope)
- Global scope
```

function setupHelp() {
  var helpText = [
    { id: 'email', help: 'Your e-mail address' },
    { id: 'name', help: 'Your full name' },
    { id: 'age', help: 'Your age (you must be over 16)' },
  ];

  for (var i = 0; i < helpText.length; i++) {
    // Culprit is the use of `var` on this line
    var item = helpText[i];

>     document.getElementById(item.id).onfocus = function () {

Reviewer: Let's use an arrow function here:
```suggestion
    document.getElementById(item.id).onfocus = () => {
```

Author: The example here is showing pre-ES6 syntax and pitfalls, so I decided to keep it as-is. Anyone who doesn't use `var` after ES6 (i.e. hopefully everyone) won't be running into this.

}

function makeHelpCallback(help) {

>   return function () {

Reviewer: ```suggestion
  return () => {
```

function setupHelp() {
  var helpText = [
    { id: 'email', help: 'Your e-mail address' },
    { id: 'name', help: 'Your full name' },
    { id: 'age', help: 'Your age (you must be over 16)' },
  ];

  for (var i = 0; i < helpText.length; i++) {
    (function () {
      var item = helpText[i];
      document.getElementById(item.id).onfocus = function () {
        showHelp(item.help);
      };
    })(); // Immediate event listener attachment with the current value of item (preserved until iteration).
  }
}

setupHelp();
```

> If you don't want to use more closures, you can use the [`let`](/en-US/docs/Web/JavaScript/Reference/Statements/let) or [`const`](/en-US/docs/Web/JavaScript/Reference/Statements/const) keyword introduced in ES6:

Reviewer: Version numbers (especially ES6, ES2015 that are 7+ years old) are useless for developers. Let's simplify:
```suggestion
If you don't want to use more closures, you can use the [`let`](/en-US/docs/Web/JavaScript/Reference/Statements/let) or [`const`](/en-US/docs/Web/JavaScript/Reference/Statements/const) keyword:
```

function MyObject(name, message) {
  this.name = name.toString();
  this.message = message.toString();

>   this.getName = function () {

Reviewer: ```suggestion
  this.getName = () => {
```

    return this.name;
  };

>   this.getMessage = function () {

Reviewer: ```suggestion
  this.getMessage = () => {
```

  this.name = name.toString();
  this.message = message.toString();
}

> MyObject.prototype.getName = function () {

Reviewer: ```suggestion
MyObject.prototype.getName = () => {
```

  return this.name;
};

> MyObject.prototype.getMessage = function () {

Reviewer: ```suggestion
MyObject.prototype.getMessage = () => {
```
